src/pk.c

```diff
@@ -30,1707 +30,2849 @@
-    byte*   encodedSig = NULL;
-#else
-    WC_RNG  tmpRNG[1];
-    byte    encodedSig[MAX_ENCODED_SIG_SZ];
-#endif
+    WOLFSSL_ENTER("wolfSSL_RSA_free");
 
-    WOLFSSL_ENTER("wolfSSL_RSA_sign_generic_padding");
+    /* Validate parameter. */
+    if (rsa == NULL) {
+        done = 1;
+    }
+    if (!done) {
+        int isZero = 0;
 
-    if (m == NULL || sigRet == NULL || sigLen == NULL || rsa == NULL) {
-        WOLFSSL_MSG("Bad function arguments");
-        return WOLFSSL_FAILURE;
+        /* Decrement reference count. */
+        if (wolfSSL_RefDec(&rsa->ref, &isZero) != 0) {
+            WOLFSSL_MSG("Couldn't lock rsa mutex");
+            done = 1;
+        }
+        else {
+            /* Nothing more to do if reference count isn't at zero. */
+            done = !isZero;
+        }
     }
-    DEBUG_SIGN_msg("Message to Sign", m, mLen);
+    if (!done) {
+        void* heap = rsa->heap;
 
-    if (rsa->inSet == 0) {
-        WOLFSSL_MSG("No RSA internal set, do it");
+        /* Dispose of allocated reference counting data. */
+        wolfSSL_RefFree(&rsa->ref);
 
-        if (SetRsaInternal(rsa) != WOLFSSL_SUCCESS) {
-            WOLFSSL_MSG("SetRsaInternal failed");
-            return WOLFSSL_FAILURE;
+    #ifdef HAVE_EX_DATA_CLEANUP_HOOKS
+        wolfSSL_CRYPTO_cleanup_ex_data(&rsa->ex_data);
+    #endif
+
+        if (rsa->internal != NULL) {
+        #if !defined(HAVE_FIPS) && !defined(HAVE_USER_RSA) && \
+            !defined(HAVE_FAST_RSA) && defined(WC_RSA_BLINDING)
+            /* Check if RNG is owned before freeing it. */
+            if (rsa->ownRng) {
+                WC_RNG* rng = ((RsaKey*)(rsa->internal))->rng;
+                if ((rng != NULL) && (rng != wolfssl_get_global_rng())) {
+                    wc_FreeRng(rng);
+                    XFREE(rng, heap, DYNAMIC_TYPE_RNG);
+                }
+                /* RNG isn't freed by wolfCrypt RSA free. */
+            }
+        #endif
+            /* Dispose of allocated data in wolfCrypt RSA key. */
+            wc_FreeRsaKey((RsaKey*)rsa->internal);
+            /* Dispose of memory for wolfCrypt RSA key. */
+            XFREE(rsa->internal, heap, DYNAMIC_TYPE_RSA);
         }
-    }
 
-    type = nid2HashSum(type);
+        /* Dispose of external representation of RSA values. */
+        wolfSSL_BN_clear_free(rsa->iqmp);
+        wolfSSL_BN_clear_free(rsa->dmq1);
+        wolfSSL_BN_clear_free(rsa->dmp1);
+        wolfSSL_BN_clear_free(rsa->q);
+        wolfSSL_BN_clear_free(rsa->p);
+        wolfSSL_BN_clear_free(rsa->d);
+        wolfSSL_BN_free(rsa->e);
+        wolfSSL_BN_free(rsa->n);
 
-    outLen = (word32)wolfSSL_BN_num_bytes(rsa->n);
+    #ifdef WC_RSA_BLINDING
+        /* Dispose of allocated data in wolfCrypt RNG. */
+        if ((rsa->rng != NULL) && (wc_FreeRng(rsa->rng) != 0)) {
+            WOLFSSL_MSG("Issue freeing rng");
+        }
+        /* Dispose of memory for wolfCrypt RNG. */
+        XFREE(rsa->rng, heap, DYNAMIC_TYPE_RNG);
```

Comment:
I can't see rsa->rng being set/used!

Reply:
It is used in `wolfSSL_EVP_PKEY_set1_RSA` and there we `wc_RsaSetRNG((RsaKey*)pkey->rsa->internal, &pkey->rng)` which means that `rsa->rng == ((RsaKey*)(rsa->internal))->rng`. This will cause a double free.

Reply:
wc_RsaSetRNG((RsaKey*)pkey->rsa->internal, &pkey->rng) is setting the wolfCrypt RSA key's rng not the wolfSSL RSA key's rng.
Don't see a need for RSA rng.